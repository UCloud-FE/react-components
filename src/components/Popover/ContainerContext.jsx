import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import createReactContext from 'create-react-context';

const ContainerContext = createReactContext();

export default ContainerContext;

export const { Consumer, Provider } = ContainerContext;

export class InhertProvider extends PureComponent {
    static propTypes = {
        value: PropTypes.object
    };
    render() {
        const { value, ...rest } = this.props;
        return (
            <Consumer>
                {context => {
                    return <Provider value={{ ...value, ...context }} {...rest} />;
                }}
            </Consumer>
        );
    }
}
