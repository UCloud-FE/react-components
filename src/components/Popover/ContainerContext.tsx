import React, { createContext, ProviderProps, useContext } from 'react';

interface Context {
    getPopupContainer?: (trigger: HTMLElement) => HTMLElement;
}
const ContainerContext = createContext<Context>({});

export default ContainerContext;

export const { Consumer, Provider } = ContainerContext;

export const InheritProvider = ({ value, ...props }: ProviderProps<Context>) => {
    const context = useContext(ContainerContext);
    return <Provider value={{ ...value, ...context }} {...props}></Provider>;
};
