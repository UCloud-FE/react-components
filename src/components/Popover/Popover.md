### 说明

-   这是 Popover 组件，弹出层功能型组件
-   使用弹出层时需要注意弹出层的容器，如果弹出层的容器为 overflow 容器且具有定位属性或在其内部或嵌套等情况（以下简称 overflow 容器），弹出内容会自动计算定位避开滚动或隐藏，防止被遮挡，但是某些情况下会导致定位的偏移，所以需要保证容器能够完整的显示弹出层，或者将容器设置到 overflow 容器的外部。
-   Popover 默认的弹出层容器为 body，可使用 getPopupContainer 来自定弹层的容器，或者使用 forwardPopupContainer 自动避开本组件库内组件的 overflow 容器（如 Card，Table 等）。
-   Modal、Drawer 较为特殊，会中断 forwardPopupContainer，从而避免找到上层非同 dom 级的容器

### 演示

#### 普通使用

```js {"codepath": "base.jsx"}
```

#### animation - 动画

```js {"codepath": "animation.jsx"}
```

#### placement - 定位

```js {"codepath": "placement.jsx"}
```

#### stretch - 尺寸自适应

```js {"codepath": "stretch.jsx"}
```

#### alignPoint - 鼠标位置定位

```js {"codepath": "point.jsx"}
```

#### visible - 控制弹出层展示(受控)

```js {"codepath": "control.jsx"}
```

#### forwardPopupContainer - 自动定位弹出层容器

```js {"codepath": "forwardPopupContainer.jsx"}
```
