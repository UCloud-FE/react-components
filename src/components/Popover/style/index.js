import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import styled from 'styled-components';
import RcTrigger from 'rc-trigger';

import config from 'config';

const { prefixCls: _prefixCls } = config;
export const prefixCls = _prefixCls + '-popover';
export const animationPrefixCls = prefixCls + '-animation';

import {
    fadeIn,
    fadeOut,
    bounceIn,
    bounceOut,
    zoomIn,
    zoomOut,
    slideUpIn,
    slideDownIn,
    slideUpOut,
    slideDownOut
} from 'src/style/animation';

const animationDuration = '0.1s';

class RcTriggerWrap extends Component {
    static propTypes = {
        className: PropTypes.string,
        popupClassName: PropTypes.string,
        trueClassName: PropTypes.string,
        triggerRef: PropTypes.func
    };
    render() {
        const { className, popupClassName, trueClassName, triggerRef, ...rest } = this.props;
        return (
            <RcTrigger
                ref={triggerRef}
                className={trueClassName}
                popupClassName={classnames(className, popupClassName)}
                {...rest}
            />
        );
    }
}

export const PopoverWrap = styled(RcTriggerWrap)`
    &.${prefixCls} {
        position: absolute;
        left: -9999px;
        top: -9999px;
        z-index: 1030;
        display: block;

        &-hidden {
            display: none;
        }
    }

    &.${prefixCls}-animation {
        &-fade {
            &-enter,
            &-appear,
            &-leave {
                animation-duration: ${animationDuration};
                animation-fill-mode: both;
            }
            &-enter,
            &-appear {
                animation-name: ${fadeIn};
            }
            &-leave {
                animation-name: ${fadeOut};
            }
        }

        &-bounce {
            &-enter,
            &-appear,
            &-leave {
                animation-duration: ${animationDuration};
                animation-fill-mode: both;
            }
            &-enter,
            &-appear {
                animation-name: ${bounceIn};
            }
            &-leave {
                animation-name: ${bounceOut};
            }
        }

        &-zoom {
            &-enter,
            &-appear,
            &-leave {
                animation-duration: ${animationDuration};
                animation-fill-mode: both;
            }
            &-enter,
            &-appear {
                animation-name: ${zoomIn};
            }
            &-leave {
                animation-name: ${zoomOut};
            }
        }

        &-slide-up {
            &-enter,
            &-appear {
                animation-duration: ${animationDuration};
                animation-fill-mode: both;
                transform-origin: 0 0;
                display: block !important;
                opacity: 0;
                animation-timing-function: cubic-bezier(0.08, 0.82, 0.17, 1);
                animation-play-state: paused;
            }
            &-leave {
                animation-duration: ${animationDuration};
                animation-fill-mode: both;
                transform-origin: 0 0;
                display: block !important;
                opacity: 1;
                animation-timing-function: cubic-bezier(0.6, 0.04, 0.98, 0.34);
                animation-play-state: paused;
            }

            &-enter-active.${prefixCls}-placement-bottomLeft,
                &-appear-active.${prefixCls}-placement-bottomLeft,
                &-enter-active.${prefixCls}-placement-bottomCenter,
                &-appear-active.${prefixCls}-placement-bottomCenter,
                &-enter-active.${prefixCls}-placement-bottomRight,
                &-appear-active.${prefixCls}-placement-bottomRight {
                animation-name: ${slideUpIn};
                animation-play-state: running;
            }
            &-enter-active.${prefixCls}-placement-topLeft,
                &-appear-active.${prefixCls}-placement-topLeft,
                &-enter-active.${prefixCls}-placement-topCenter,
                &-appear-active.${prefixCls}-placement-topCenter,
                &-enter-active.${prefixCls}-placement-topRight,
                &-appear-active.${prefixCls}-placement-topRight {
                animation-name: ${slideDownIn};
                animation-play-state: running;
            }
            &-leave-active.${prefixCls}-placement-bottomLeft,
                &-leave-active.${prefixCls}-placement-bottomCenter,
                &-leave-active.${prefixCls}-placement-bottomRight {
                animation-name: ${slideUpOut};
                animation-play-state: running;
            }
            &-leave-active.${prefixCls}-placement-topLeft,
                &-leave-active.${prefixCls}-placement-topCenter,
                &-leave-active.${prefixCls}-placement-topRight {
                animation-name: ${slideDownOut};
                animation-play-state: running;
            }
        }
    }
`;
