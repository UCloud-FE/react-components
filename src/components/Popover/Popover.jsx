import React, { Component } from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import { polyfill } from 'react-lifecycles-compat';

import placements from './placements';
import { prefixCls, animationPrefixCls, PopoverWrap } from './style';

const Animation = ['fade', 'zoom', 'bounce', 'slide-up'];
const Trigger = ['hover', 'focus', 'click', 'contextMenu'];
const Placement = Object.keys(placements);

class Popover extends Component {
    static propTypes = {
        /** 受控，控制弹出层展示 */
        visible: PropTypes.bool,
        /** 非受控，是否默认展示弹出层 */
        defaultVisible: PropTypes.bool,
        /** 弹出层显示隐藏时触发 */
        onVisibleChange: PropTypes.func,
        /** 如何触发弹出层，focus需要注意被包裹元素必须能触发focus事件，如链接、按钮、input等 */
        trigger: PropTypes.arrayOf(PropTypes.oneOf(Trigger)),
        /** 根据鼠标位置定位 */
        alignPoint: PropTypes.bool,
        /** @ignore */
        showAction: PropTypes.array,
        /** @ignore */
        hideAction: PropTypes.array,
        /** 位置 */
        placement: PropTypes.oneOf(Placement),
        /**
         * @ignore
         * 自定义定位配置
         */
        builtinPlacements: PropTypes.object,
        /** 自定义定位 */
        align: PropTypes.object,
        /** 尺寸自适应，'width', 'minWidth', 'height', 'minHeight' 或混合使用 */
        stretch: PropTypes.arrayOf(PropTypes.oneOf(['width', 'minWidth', 'height', 'minHeight'])),
        /** 弹出层内容 */
        popup: PropTypes.node,
        /** 弹出层的类名 */
        popupClassName: PropTypes.string,
        /** 弹出层的样式 */
        popupStyle: PropTypes.object,
        /** 弹出层的z-index */
        zIndex: PropTypes.number,
        /** 自定义弹出层容器 */
        getPopupContainer: PropTypes.func,
        /**
         * @ignore
         * 自定义类名前缀
         */
        prefixCls: PropTypes.string,
        /** @ignore */
        children: PropTypes.any,
        /** 动画名称，slide-up只支持上下方向的弹窗 */
        animation: PropTypes.oneOf(Animation),
        /**
         * @ignore
         * 自定义动画
         */
        transitionName: PropTypes.string,
        /**
         * 更新时强制重新定位
         * @ignore
         */
        forceAlignWhenUpdate: PropTypes.bool,
        /** @ignore */
        className: PropTypes.string
    };

    static defaultProps = {
        trigger: ['hover'],
        showAction: [],
        hideAction: [],
        popupClassName: '',
        animation: 'fade',
        popupStyle: {},
        defaultVisible: false,
        onVisibleChange() {},
        placement: 'bottomLeft',
        builtinPlacements: placements,
        prefixCls: prefixCls
    };

    constructor(props) {
        super(props);
        if ('visible' in props) {
            this.state = {
                visible: props.visible
            };
        } else {
            this.state = {
                visible: props.defaultVisible
            };
        }
    }

    componentDidMount = () => {
        document.addEventListener('scroll', this.forceAlign, true);
    };
    componentWillUnmount = () => {
        document.removeEventListener('scroll', this.forceAlign, true);
    };
    forceAlign = _.debounce(() => {
        this.trigger && this.trigger.forcePopupAlign();
    }, 100);

    componentDidUpdate = () => {
        this.props.forceAlignWhenUpdate && this.forceAlign();
    };

    onVisibleChange = visible => {
        const props = this.props;
        if (!('visible' in props)) {
            this.setState({
                visible
            });
        }
        props.onVisibleChange(visible);
    };

    static getDerivedStateFromProps(nextProps) {
        if ('visible' in nextProps) {
            return {
                visible: nextProps.visible
            };
        }
        return null;
    }

    getPopup() {
        const { popup } = this.props;
        return popup;
    }

    getPopupDomNode() {
        return this.trigger.getPopupDomNode();
    }

    saveTrigger = node => {
        this.trigger = node;
    };

    render() {
        const {
            children,
            transitionName,
            animation,
            align,
            placement,
            trigger,
            stretch = [],
            className,
            ...rest
        } = this.props;
        const popup = this.getPopup();
        return (
            <PopoverWrap
                {...rest}
                triggerRef={this.saveTrigger}
                action={trigger}
                popupPlacement={placement}
                popupAlign={align}
                popupTransitionName={transitionName || animation ? animationPrefixCls + '-' + animation : null}
                popupVisible={popup == null ? false : this.state.visible}
                popup={popup == null ? <div /> : popup}
                onPopupVisibleChange={this.onVisibleChange}
                stretch={stretch.join('')}
                trueClassName={className}
            >
                {children}
            </PopoverWrap>
        );
    }
}

Object.assign(Popover, {
    Animation,
    Trigger,
    Placement
});

polyfill(Popover);
export default Popover;

export { prefixCls };
