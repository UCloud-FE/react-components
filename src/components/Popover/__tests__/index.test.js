import React from 'react';
import { mount } from 'enzyme';
import { renderToJson } from 'enzyme-to-json';

import Popover from 'src/components/Popover';

jest.unmock('rc-trigger');
describe('Popover', () => {
    test('popover align', () => {
        const wrapper = mount(
            <Popover popup={<div>popup</div>} forceAlignWhenUpdate getPopupContainer={triggerNode => triggerNode}>
                <div className="content">content</div>
            </Popover>
        );
        expect(renderToJson(wrapper.render())).toMatchSnapshot();
        wrapper.setProps({
            visible: true
        });
        expect(renderToJson(wrapper.render())).toMatchSnapshot();

        wrapper.instance().getPopupDomNode();
        wrapper.instance().forceAlign();
    });

    test('empty', () => {
        const wrapper = mount(
            <Popover>
                <div className="content">content</div>
            </Popover>
        );
        expect(renderToJson(wrapper.render())).toMatchSnapshot();
    });
});
