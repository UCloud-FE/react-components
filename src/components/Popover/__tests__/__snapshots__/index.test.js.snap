// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover empty 1`] = `
<div
  class="content"
>
  content
</div>
`;

exports[`Popover popover align 1`] = `
<div
  class="content"
>
  content
</div>
`;

exports[`Popover popover align 2`] = `
<div
  class="content"
  style=""
>
  content
  <div
    style="position: absolute; top: 0px; left: 0px; width: 100%;"
  >
    <div>
      <div
        class="uc-fe-popover style__PopoverWrap-cfqvNq iMAKfT uc-fe-popover-placement-bottomLeft uc-fe-popover-animation-fade-appear"
        style="left: -999px; top: -994px;"
      >
        <div>
          popup
        </div>
      </div>
    </div>
  </div>
</div>
`;
